Implement titlecase iterator

choznerol opened the pull request:
Fix #65

> Unlike Lowercase and Uppercase, `char::to_titlecase` is not defined in `std`, so we'll have to generate mappings for the titlecase transforms.
>
> One way we might do that is to roll some Ruby scripts like focaccia does for generating case folding mappings:
>
> - https://github.com/artichoke/focaccia/blob/ed00eb45e6d1a38dda145cfae2b64c215119a2fe/scripts/gen_case_lookups.rb
> - https://github.com/artichoke/focaccia/blob/ed00eb45e6d1a38dda145cfae2b64c215119a2fe/Rakefile#L92-L104
>
> Another way we might tackle this is with codegen from crates in the Rust ecosystem:
>
> - https://crates.io/crates/ucd-generate

The `char::to_titlecase` was removed from `std` in https://github.com/rust-lang/rust/issues/26555.
Before actually start implementing either of the approach using `UnicodeData.txt`, I found a crate [`unicode_titlecase`] implementing `char::to_titlecase()` was created six months after #65 was opened.
If using this crate directly matches the project goal, the Titlecase implementation would be pretty similar to Uppercase and Lowercase (current approach); Otherwise, we'll have to do some mapping with `UnicodeData.txt` and `SpecialCasing.txt` similar to [how the create does](https://github.com/Teh-Bobo/unicode-title-case/blob/e7955f0157480621445a029920748b91ded388d4/build.rs#L12-L13).

<!--

I did a quick check with Ruby `String#capitalize`:

```shell
ruby -e 'puts ["ß", "ǅ", "İ", "İ", "ΈΤΟΣ", "ZȺȾ"].map{|c| "#{c} -> #{c.capitalize}\n#{c.codepoints} -> #{c.capitalize.codepoints}\n\n"}' 

ß -> Ss
[223] -> [83, 115]

ǅ -> ǅ
[453] -> [453]

İ -> İ
[304] -> [304]

İ -> İ
[304] -> [304]

ΈΤΟΣ -> Έτοσ
[904, 932, 927, 931] -> [904, 964, 959, 963]

ZȺȾ -> Zⱥⱦ
[90, 570, 574] -> [90, 11365, 11366]
```

I copied and modified `case_map_to_(two|three)_chars` mainly from `tests::uppercase::full::tests`. The expected result matches with Ruby's `String#capitalize` returns. However, I'm not sure how those test cases were chosen initially, and if more/fewer cases are suitable for titlecase. For example, should I just remove the `ᾂ` test case considering it doesn't map to map to two chars here?

```shell
$  ruby -e 'puts ["case", "և", "ẙ", "ᾂ", "ﬗ", "ﬃ", "ὖ", "ῷ"].map{|c| "#{c} -> #{c.upcase} / #{c.capitalize} \n #{c.codepoints} -> #{c.upcase.codepoint
s} / #{c.capitalize.codepoints}\n\n"}'
case -> CASE / Case 
 [99, 97, 115, 101] -> [67, 65, 83, 69] / [67, 97, 115, 101]

և -> ԵՒ / Եւ 
 [1415] -> [1333, 1362] / [1333, 1410]

ẙ -> Y̊ / Y̊ 
 [7833] -> [89, 778] / [89, 778]

ᾂ -> ἊΙ / ᾊ 
 [8066] -> [7946, 921] / [8074]

ﬗ -> ՄԽ / Մխ 
 [64279] -> [1348, 1341] / [1348, 1389]

ﬃ -> FFI / Ffi 
 [64259] -> [70, 70, 73] / [70, 102, 105]

ὖ -> Υ̓͂ / Υ̓͂ 
 [8022] -> [933, 787, 834] / [933, 787, 834]

ῷ -> Ω͂Ι / ῼ͂ 
 [8183] -> [937, 834, 921] / [937, 834, 837]
```

-->

[`unicode_titlecase`]: https://docs.rs/unicode_titlecase/2.0.0/unicode_titlecase/#traits

choznerol committed:
Implement `TitlecaseMode` and ASCII iterator

Partly fix #65

choznerol committed:
Implement `full::Titlecase` with `unicode_titlecase`

https://github.com/Teh-Bobo/unicode-title-case

choznerol commented on src/titlecase/full.rs:
```diff
@@ -0,0 +1,416 @@
+    fn utf8() {
+        let s = "ß".as_bytes();
+        let iter = Titlecase::from(s);
+        assert_eq!(
+            iter.collect::<Vec<_>>().as_bstr(),
+            "Ss".as_bytes().as_bstr()
+        );
+
+        let s = "Αύριο".as_bytes();
+        let iter = Titlecase::from(s);
+        assert_eq!(
+            iter.collect::<Vec<_>>().as_bstr(),
+            "Αύριο".as_bytes().as_bstr()
+        );
+
+        let s = "Έτος".as_bytes();
+        let iter = Titlecase::from(s);
+        assert_eq!(
+            iter.collect::<Vec<_>>().as_bstr(),
+            "Έτος".as_bytes().as_bstr()
+        );
+
+        // two-byte characters
+        // https://github.com/minimaxir/big-list-of-naughty-strings/blob/894882e7/blns.txt#L198-L200
+        let s = "𐐜 𐐔𐐇𐐝𐐀𐐡𐐇𐐓 𐐙𐐊𐐡𐐝𐐓/𐐝𐐇𐐗𐐊𐐤𐐔 𐐒𐐋𐐗 𐐒𐐌 𐐜 𐐡𐐀𐐖𐐇𐐤𐐓𐐝 𐐱𐑂 𐑄 𐐔𐐇𐐝𐐀𐐡𐐇𐐓 𐐏𐐆𐐅𐐤𐐆𐐚𐐊𐐡𐐝𐐆𐐓𐐆".as_bytes();
+        let iter = Titlecase::from(s);
+        assert_eq!(
+            iter.collect::<Vec<_>>().as_bstr(),
+            "𐐜 𐐼𐐯𐑅𐐨𐑉𐐯𐐻 𐑁𐐲𐑉𐑅𐐻/𐑅𐐯𐐿𐐲𐑌𐐼 𐐺𐐳𐐿 𐐺𐐴 𐑄 𐑉𐐨𐐾𐐯𐑌𐐻𐑅 𐐱𐑂 𐑄 𐐼𐐯𐑅𐐨𐑉𐐯𐐻 𐐷𐐮𐐭𐑌𐐮𐑂𐐲𐑉𐑅𐐮𐐻𐐮"
+                .as_bytes()
+                .as_bstr()
+        );
+
+        // Change length when titlecased
+        // https://github.com/minimaxir/big-list-of-naughty-strings/blob/894882e7/blns.txt#L226-L232
+        let s = "ZȺȾ".as_bytes();
+        let iter = Titlecase::from(s);
+        assert_eq!(
+            iter.collect::<Vec<_>>().as_bstr(),
+            "Zⱥⱦ".as_bytes().as_bstr()
```
I did a quick check with Ruby `String#capitalize`:

```shell
ruby -e 'puts ["ß", "ǅ", "İ", "İ", "ΈΤΟΣ", "ZȺȾ"].map{|c| "#{c} -> #{c.capitalize}\n#{c.codepoints} -> #{c.capitalize.codepoints}\n\n"}' 

ß -> Ss
[223] -> [83, 115]

ǅ -> ǅ
[453] -> [453]

İ -> İ
[304] -> [304]

İ -> İ
[304] -> [304]

ΈΤΟΣ -> Έτοσ
[904, 932, 927, 931] -> [904, 964, 959, 963]

ZȺȾ -> Zⱥⱦ
[90, 570, 574] -> [90, 11365, 11366]
```

choznerol commented on src/titlecase/full.rs:
```diff
@@ -0,0 +1,416 @@
+    }
+
+    #[test]
+    fn case_map_to_two_chars() {
+        let s = "և".as_bytes();
+        let iter = Titlecase::from(s);
+        assert_eq!(
+            iter.collect::<Vec<_>>().as_bstr(),
+            "Եւ".as_bytes().as_bstr()
+        );
+
+        let s = "ẙ".as_bytes();
+        let iter = Titlecase::from(s);
+        assert_eq!(
+            iter.collect::<Vec<_>>().as_bstr(),
+            "Y\u{30a}".as_bytes().as_bstr()
+        );
+
+        let s = "ᾂ".as_bytes();
+        let iter = Titlecase::from(s);
+        assert_eq!(iter.collect::<Vec<_>>().as_bstr(), "ᾊ".as_bytes().as_bstr());
+
+        let s = "ﬗ".as_bytes();
+        let iter = Titlecase::from(s);
+        assert_eq!(
+            iter.collect::<Vec<_>>().as_bstr(),
+            "Մխ".as_bytes().as_bstr()
+        );
+    }
+
+    #[test]
+    fn case_map_to_three_chars() {
```
I copied and modified `case_map_to_(two|three)_chars` mainly from `tests::uppercase::full::tests`. The expected result matches with Ruby's `String#capitalize` returns. However, I'm not sure how those test cases were chosen initially, and if more/fewer cases are suitable for titlecase. For example, should I just remove the `ᾂ` test case considering it doesn't map to map to two chars here?

```shell
$  ruby -e 'puts ["case", "և", "ẙ", "ᾂ", "ﬗ", "ﬃ", "ὖ", "ῷ"].map{|c| "#{c} -> #{c.upcase} / #{c.capitalize} \n #{c.codepoints} -> #{c.upcase.codepoint
s} / #{c.capitalize.codepoints}\n\n"}'
case -> CASE / Case 
 [99, 97, 115, 101] -> [67, 65, 83, 69] / [67, 97, 115, 101]

և -> ԵՒ / Եւ 
 [1415] -> [1333, 1362] / [1333, 1410]

ẙ -> Y̊ / Y̊ 
 [7833] -> [89, 778] / [89, 778]

ᾂ -> ἊΙ / ᾊ 
 [8066] -> [7946, 921] / [8074]

ﬗ -> ՄԽ / Մխ 
 [64279] -> [1348, 1341] / [1348, 1389]

ﬃ -> FFI / Ffi 
 [64259] -> [70, 70, 73] / [70, 102, 105]

ὖ -> Υ̓͂ / Υ̓͂ 
 [8022] -> [933, 787, 834] / [933, 787, 834]

ῷ -> Ω͂Ι / ῼ͂ 
 [8183] -> [937, 834, 921] / [937, 834, 837]
```

b-n replied:
IMO, I would move the test cases that don't look right to the correct test case. e.g. moving:

```
ᾂ -> ἊΙ / ᾊ [8066] -> [7946, 921] / [8074]
```

to the `case_map_to_two_chars` based on what you've got here.

I suspect there might be some unicode [de]normalization going on here since U+1F82 can be denormalized into 4 codepoints: https://www.compart.com/en/unicode/U+1F82, and thus I suspect there is some other combinations around for this.

choznerol replied:
> I would move the test cases that don't look right to the correct test case.

Ah, that make sense. Fixed in 09e6fe5df25cdd0d37b37bf7c8396d7a9487bcbf.

> I suspect there might be some unicode [de]normalization going on here since U+1F82 can be denormalized into 4 codepoints: https://www.compart.com/en/unicode/U+1F82, and thus I suspect there is some other combinations around for this.

I'm not sure if Unicode composition is related to case mapping 🤔 IIUC titlecasing is just a simple one directional mapping similar to uppercasing and lowercasing. 

https://www.compart.com/en/unicode/U+1F82 only shows its `Uppercase Character`, but https://codepoints.net/U+1F82 does show all its three kinds of case mapping.

lopopolo commented:
> If using this crate directly matches the project goal

This crate seems to be ~unused in the ecosystem from an author I don't recognize. The repo itself has very few stars. I don't think I'd be comfortable taking a dependency on this crate if we don't need to. Can you explore the mapping approach instead?

lopopolo commented on Cargo.toml:
```diff
@@ -24,6 +24,7 @@ alloc = []
 
 [dependencies]
 bstr = { version = "1.0.1", default-features = false }
+unicode_titlecase = "2.0.0"
```
I'm not comfortable using this dep

choznerol commented:
> > If using this crate directly matches the project goal
> 
> This crate seems to be ~unused in the ecosystem from an author I don't recognize. The repo itself has very few stars. I don't think I'd be comfortable taking a dependency on this crate if we don't need to. Can you explore the mapping approach instead?

Sure, let me explore the mapping approach 👍

choznerol committed:
Fix clippy warning

choznerol committed:
`bundle exec rake unicode:update` to fetch unicode data

choznerol committed:
`bundle exec rake unicode:build` to run ucd-generate

choznerol committed:
Implement char::to_titlecase()

choznerol commented on Rakefile:
```diff
@@ -92,6 +92,21 @@ end
+    end
+
+    dest = "./generated/case_mapping.rs"
+    if system("ucd-generate case-mapping #{ucd_dir} --flat-table > #{dest}")
```
We only need `TITLE`, and it's surprisingly hard to ignore or autocorrect the unused `UPPER` and `LOWER` with the `dead_code` rule.

I'm trying to make it possible to generate only `TITLE` with a new `ucd-generate` flag  in https://github.com/BurntSushi/ucd-generate/pull/55.

```suggestion
    if system("ucd-generate case-mapping #{ucd_dir} --flat-table --include TITLE > #{dest}")
```

choznerol committed:
Use ucd-generate#55 for now

choznerol committed:
Use ucd-generate 0.3.0

choznerol committed:
Improve coverage from 92% to 97% to pass CI

https://github.com/artichoke/roe/actions/runs/5493052390/jobs/10010965555?pr=139

 On branch trunk:
coveragePercent = 93.35
linesCovered = 1460
linesMissed = 104
linesTotal = 1564

On PR artichoke/roe#139:
coveragePercent = 92.17
linesCovered = 2212
linesMissed = 188
linesTotal = 2400

choznerol commented on src/unicode/std_case_mapping_iter.rs:
```diff
@@ -0,0 +1,115 @@
+// This file contain code extracted from std::char::CaseMappingIter.
```
Unfortunately [`std::char::CaseMappingIter`](https://doc.rust-lang.org/src/core/char/mod.rs.html#445) was not export, so I end up extracting part of them in this file.

b-n commented on Rakefile:
```diff
@@ -89,6 +89,37 @@ task :test do
+      exit(1)
+    end
+
+    unless system('git diff --exit-code')
```
I'm wondering if something like this to be a bit more specific to the changes, however I see the later cargo clippy, and couldn't find a way to make that run specifically in the directory, so maybe a full dirty check is just better. Just leaving the suggestion in case other things come to mind.

```suggestion
    unless system('git diff --exit-code -- "generated/**/*"')
```

choznerol replied:
> couldn't find a way to make that run specifically in the directory

So do I. Seem like the only way is to have a separate package (`cargo clippy -p ...`).

The motivation for this check was to prevent the later `cargo clippy --fix` from modifying unstaged work unexpectedly. In contrast, I'm not worried too much about modifying `generated/**/*`, since they can be easily re-generated.

b-n commented on Rakefile:
```diff
@@ -89,6 +89,37 @@ task :test do
 end
 
+namespace :unicode do
+  ucd_dir = 'generated/ucd'
```
```suggestion
  ucd_dir = Pathname.pwd.join('generated', 'ucd')
```

likely will need `require 'pathname'` at the top too

choznerol replied:
Thanks. Fixed in 314f5b2.
By the way, I end up using relative path in some places to avoid: <img width="965" alt="image" src="https://github.com/artichoke/roe/assets/12410942/b8327c83-ed21-42c0-94a3-7dcc98546079">

b-n commented on Rakefile:
```diff
@@ -89,6 +89,37 @@ task :test do
+      exit(1)
+    end
+
+    filename = './generated/case_mapping.rs'
```
```suggestion
    filename = Pathname.pwd.join('generated', 'case_mapping.rs')
```

choznerol replied:
Fixed in 314f5b2315b660f588b7e12944bb01ae9fd90f19 too

b-n commented on Rakefile:
```diff
@@ -89,6 +89,37 @@ task :test do
+    %w[UnicodeData.txt SpecialCasing.txt PropList.txt].each do |filename|
+      uri = "https://www.unicode.org/Public/UCD/latest/ucd/#{filename}"
+      URI.parse(uri).open do |data|
+        IO.copy_stream(data, "#{ucd_dir}/#{filename}")
```
```suggestion
        IO.copy_stream(data, "#{Pathname.pwd.join(ucd_dir, filename)}")
```

choznerol replied:
Fixed in 314f5b2315b660f588b7e12944bb01ae9fd90f19 too

b-n commented on generated/ucd/README.md:
```diff
@@ -0,0 +1,3 @@
+Do not edit `generated/ucd/*.txt`. They were automatically generated by:
```
```suggestion
Do not edit `generated/ucd/*.txt`. The files are automatically generated by:
```

choznerol replied:
Thanks. Fixed in 01d0ac0.

b-n commented on generated/ucd/README.md:
I think this file should be in `generated/README.md` instead of this sub folder.

choznerol replied:
Fixed in 01d0ac0 too.

b-n commented on src/unicode.rs:
```diff
@@ -0,0 +1,5 @@
+pub mod titlecase;
+mod ucd_generated_case_mapping;
+
+pub use titlecase::{to_titlecase, Titlecase, ToTitlecase};
```
```suggestion
#[cfg(feature = "alloc")]
pub use titlecase::{to_titlecase, Titlecase, ToTitlecase};
```

choznerol replied:
But `Vec` isn't being used in `src/unicode/*` except in `tests`. I assume `cargo test --no-default-features` will catch this if `Vec` is actually used.
<details><summary>example</summary>
<img width="840" alt="image" src="https://github.com/artichoke/roe/assets/12410942/5fcff985-de70-4a1f-aac8-748f3bbfdb9e">
</details>

b-n replied:
Oh wow, good call! I did the naive thing of checking the ascii impl, and saw `to_ascii_lowercase` had the feature alloc around it.

The good news, after changing around the module structure, I now see what's going on I think.

It looks like `roe` currently exports not only the titlecase/lowercase/uppercase iterators, but also encoding specific functions - specifically things like `ascii::lowercase::to_ascii_lowercase`. I assume these functions are to be consumed by other Ruby string functions as short circuits for perf reasons (instead of slice => iterator => vec), and keeping all "case folding logic" in one crate.

This makes me wonder though, should `unicode::*` be exported or not. If yes, I think it would make sense to make `unicode::titlecase::to_titlecase` to be consistent with `ascii::titlecase::to_titlecase`, and thus to take a slice, and return a newly allocated `Vec<u8>`.

My preference:
- If we export `ascii::*` related methods, then we should also export `unicode::*` related things too
- `ascii::*` and `unicode::*` should take the same sort of parameters in this case

On "whether we should export `ascii::*`", I think it makes sense, and thus use `roe::ascii::*` instead of `std::str::*` in `artichoke-backend` and friends. @lopopolo thoughts?

b-n replied:
Note: if we do want to export `unicode::*`, then I would say for this PR we just don't export `unicode::*` for now, and then in the next PR do the work to bring some of the unicode functions into the `unicode::*` module etc.

b-n commented on src/unicode.rs:
I'm wondering why the original structure didn't have `src/ascii.rs` in `src/ascii/mod.rs`. How to you feel about moving this to `src/unicode/mod.rs` and also moving `src/ascii.rs` to `src/ascii/mod.rs` ?

choznerol replied:
I didn't find a difference between these two ways, but `src/ascii.rs -> src/ascii/mod.rs` also feel more intuitive to me too. Fix in 2fca461.

b-n commented on src/unicode/std_case_mapping_iter.rs:
```diff
@@ -0,0 +1,136 @@
+// This file contain code extracted from std::char::CaseMappingIter.
```
```suggestion
// This file contains code extracted from std::char::CaseMappingIter.
```

choznerol replied:
Fixed in 01d0ac0 too.

b-n commented on src/titlecase.rs:
````diff
@@ -0,0 +1,273 @@
+    /// let mut expected = "Αύριο".to_string().into_bytes();
+    /// expected.extend(b"\xFF\xFE");
+    /// assert_eq!(titlecase.collect::<Vec<_>>(), expected);
+    /// ```
````
Another example of the transformation into multiple codepoints I think would be really nice to see. e.g. the expansion of the ffi ligature and it's specific quirks since those are the crazy ones.

choznerol replied:
Sure. Added in 2953a24.

b-n commented on src/titlecase.rs:
```diff
@@ -0,0 +1,273 @@
+use core::iter::FusedIterator;
```
This file module is very nice! It lines up really nicely with the lower/uppercase modules!

b-n commented on src/uppercase/full.rs:
```diff
@@ -390,4 +390,14 @@ mod tests {
     }
+
+    #[test]
+    fn test_fmt() {
```
:rocket:

b-n commented on src/lowercase/full.rs:
```diff
@@ -359,4 +359,14 @@ mod tests {
     }
+
+    #[test]
+    fn test_fmt() {
```
:rocket:

b-n commented on src/titlecase/ascii.rs:
```diff
@@ -0,0 +1,297 @@
+    type Item = u8;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let (&byte, remainder) = self.slice.split_first()?;
```
This is absolutely genius way of doing this. I was thinking whether the self.first could have been made into a guard clause, but you'll always need the next char, so this is nice and efficient!

choznerol replied:
I recall this one was ChatGPT's idea 😆

b-n replied:
Ahh, the AI is here to take our jobs :sweat_smile:. You win this time ChatGPT!

b-n commented on src/titlecase/ascii.rs:
```diff
@@ -0,0 +1,297 @@
+        assert_eq!(
+            iter.collect::<Vec<_>>().as_bstr(),
+            "ZȺȾ".as_bytes().as_bstr()
+        );
```
Thinking whether this should be checking the byte length, and validating that it shouldn't change rather than the output characters (as per the comment above it)

choznerol replied:
You're right. In fact, I realized `zȺȾ -> ZȺȾ` doesn't test much about the changing bytes length 😅. 
Fixed in 3e5fcee.

b-n commented on src/lib.rs:
```diff
@@ -472,3 +609,109 @@ macro_rules! readme {
 readme!();
+
+#[cfg(test)]
+mod tests {
```
:heart:  new tests!

b-n commented on src/titlecase/full.rs:
```diff
@@ -0,0 +1,428 @@
+    next_bytes: [u8; 4],
+    next_range: Range<usize>,
+    case_iter: Option<ToCase>,
+    beginning: bool,
```
`first` instead of `beginning` to be consistent with the ascii implementation?

choznerol replied:
Thanks I didn't notice that one! Fixed in 01d0ac0.

b-n reviewed (changes requested):
A quick round of review on the "easy" parts :sweat_smile:. I'll come back and look at the more chunky `src/unicode/titlecase.rs` and `src/titlecase/full.rs` though, just wanted to give you some small nits quickly.

I should say, in general it looks really good!

choznerol commented on Rakefile:
```diff
@@ -89,6 +89,37 @@ task :test do
+
+  desc 'Rebuild Rust generated Rust sources from Unicode data'
+  task :build do
+    unless system('ucd-generate --version')
+      puts 'Please install ucd-generate > 0.3.0 first'
+      exit(1)
+    end
```
An older version of `ucd-generate` won't be caught by this. Let me actually check the installed version to be >= 0.3.0 too.

<!--


<img width="770" alt="error ucd-generate not installed" src="https://github.com/artichoke/roe/assets/12410942/109cf1d8-102e-46b3-95e8-c6c9dc072b86">

<img alt="error please upgrade ucd-generate" src="https://github.com/artichoke/roe/assets/12410942/81ccc30a-4794-4393-a753-cb99ff8a7e64">

-->

choznerol replied:
Fixed in 2bbedfb.

Error for `ucd-generate` not installed:
> <img width="770" alt="error ucd-generate not installed" src="https://github.com/artichoke/roe/assets/12410942/109cf1d8-102e-46b3-95e8-c6c9dc072b86">

Error for older version of `ucd-generate`:
> <img alt="error please upgrade ucd-generate" src="https://github.com/artichoke/roe/assets/12410942/81ccc30a-4794-4393-a753-cb99ff8a7e64">

choznerol committed:
Check the version of installed ucd-generate >= 0.3.0 (https://github.com/artichoke/roe/pull/139#discussion_r1299056371)

choznerol committed:
Use Filename instead of hard-coding path

https://github.com/artichoke/roe/pull/139#discussion_r1288643484
https://github.com/artichoke/roe/pull/139#discussion_r1288643619
https://github.com/artichoke/roe/pull/139#discussion_r1288646885

Co-authored-by: Ben Naylor <nayben@gmail.com>

choznerol committed:
Fix typo and improve naming

https://github.com/artichoke/roe/pull/139#discussion_r1288653031
https://github.com/artichoke/roe/pull/139#discussion_r1288650358
https://github.com/artichoke/roe/pull/139#discussion_r1288995258
https://github.com/artichoke/roe/pull/139#discussion_r1288961349

Co-authored-by: Ben Naylor <nayben@gmail.com>

choznerol committed:
Improve test case for Ⱥ/ⱥ and Ⱦ/ⱦ (https://github.com/artichoke/roe/pull/139#discussion_r1288986467)

The byte length of Ⱥ and ⱥ (or Ⱦ and ⱦ) is different. For ascii test cases, we can assert that they are skipped so byte length remain equal; for UTF-8 test cases, we can assert that the case change also results in change in byte length.

choznerol committed:
Test ﬃ titlecase (https://github.com/artichoke/roe/pull/139#discussion_r1288970270)

choznerol committed:
Use mod.rs for project structure (https://github.com/artichoke/roe/pull/139#discussion_r1288959639)

b-n commented on src/titlecase/full.rs:
```diff
@@ -0,0 +1,426 @@
+                };
+                let ch = case_iter
+                    .next()
+                    .expect("ToTitlecase or ToLowercase yields at lteast one char");
```
```suggestion
                    .expect("ToTitlecase or ToLowercase yields at least one char");
```

choznerol replied:
Fixed in 43f13a7a6332fce45e86b348a94f1d45f439a57b. Tks!

b-n commented on src/unicode/std_case_mapping_iter.rs:
```diff
@@ -0,0 +1,136 @@
+// This file contains code extracted from std::char::CaseMappingIter.
```
I hadn't checked through the need for this until just now, and I think I understand! Very cool.

So, to correct my understanding here, we need this since the iterator of a unicode titlecase because when we case map, we could end up with more unicode code points than what came in (e.g. 1 code point uppercase might be 3 lowercase chars). I'm trying to understand by reading the code through, but my brain is mush, and maybe it's easier for you to explain it to me - but my question is: Is it possible to wrap `core::char::ToLowerCase` and `core::char::ToUpperCase` iterators instead of bringing this in? It's entirely possible I'm missing something important here :sweat_smile:

choznerol replied:
> we could end up with more unicode code points than what came in (e.g. 1 code point uppercase might be 3 lowercase chars).

Exactly 💯 at least that's also how I understand it!

> Is it possible to wrap `core::char::ToLowerCase` and `core::char::ToUpperCase` 

That's actually a very good idea! However, I gave it a try in https://github.com/artichoke/roe/commit/ada819cc8b4efca9e1af66b5c7da9875c75374f2#diff-1beb18474b7675a5d085ca3787c8cd13badcc4660db8aa63f6febb96ec71faa0L74-R84 but realized there wasn't a public API to construct a `ToLowercase`.

It appears [`char::to_lowercase()`](https://doc.rust-lang.org/stable/std/primitive.char.html#method.to_lowercase) is the only way to obtain a `ToLowercase`.  There isn't a method to mutate the internal of a [`ToLowercase`](https://doc.rust-lang.org/stable/std/char/struct.ToLowercase.html) either 😞.

Thanks for the idea still, and please let me know if you can think of any way that I can try. I'd really prefer to not copy `CaseMappingIter` if possible 😂.

lopopolo replied:
Title case is a distinct operation from upper and lowercase.

This character is the example I know:

> Dž (titlecase form; all-capitals form DŽ, lowercase dž)

b-n reviewed (dismissed):
Found another tiny typo, but TBH it's looking really close to being there. I finally dived in and looked at the difference between `titlecase::full` and `lowercase::full`, and I see what you've done - very cool! I don't have many comments because of how close we are I think :smile:

choznerol committed:
Fix typo

https://github.com/artichoke/roe/pull/139#discussion_r1302864733

Co-authored-by: b-n <nayben@gmail.com>

choznerol committed:
Move ᾂ to appropriate test case (https://github.com/artichoke/roe/pull/139#discussion_r1302872742)

choznerol reviewed (commented):
Thanks for the detailed review!

lopopolo commented on src/titlecase/ascii.rs:
```diff
@@ -0,0 +1,295 @@
+    fn next_back(&mut self) -> Option<Self::Item> {
+        let (&byte, remainder) = self.slice.split_last()?;
+        self.slice = remainder;
+        if remainder.is_empty() {
```
There is a bug here with this sequence:

1. Create iterator
2. Call iter.next() to consume the title case / first case.
3. Call iter.next_back() until the iterator is consumed.

This will result in 2 uppercase ASCII chars being emitted.

choznerol replied:
Thanks! Reproduced in 42e1f79 and fixed in 6ffc252.

choznerol committed:
Test DoubleEndedIterator

choznerol committed:
Fix https://github.com/artichoke/roe/pull/139#discussion_r1323298582

lopopolo committed:
Flesh out Ǆ tests for all casing combinations

lopopolo committed:
Mark ucd state machines and tables as generated

lopopolo committed:
Add title to generated README

lopopolo reviewed (approved):
Added a couple of fixups. this looks great! Sorry this took so long to get approved.

Once this is merged, I'll prep a new release.

lopopolo commented:
@b-n it looks like your :x: is only for a typo that @choznerol fixed. I'm gonna dismiss your review and get this merged

b-n commented:
@lopopolo Ah yes, all good! I should have cleared my review 😅

choznerol commented:
> Added a couple of fixups. this looks great! Sorry this took so long to get approved.
> 
> Once this is merged, I'll prep a new release.

No worry. Thank you guys for the review!
